app/javascript/packages/webauthn/webauthn-verify-button-element.ts

```diff
@@ -0,0 +1,72 @@
+import verifyWebauthnDevice from './verify-webauthn-device';
+import type { VerifyCredentialDescriptor } from './verify-webauthn-device';
+
+export interface WebauthnVerifyButtonDataset extends DOMStringMap {
+  credentials: string;
+  userChallenge: string;
+}
+
+class WebauthnVerifyButtonElement extends HTMLElement {
+  dataset: WebauthnVerifyButtonDataset;
+
+  connectedCallback() {
+    this.bindEvents();
+  }
+
+  get button(): HTMLButtonElement {
+    return this.querySelector('.webauthn-verify-button__button')!;
+  }
+
+  get spinner(): HTMLElement {
+    return this.querySelector('.webauthn-verify-button__spinner')!;
+  }
+
+  get credentials(): VerifyCredentialDescriptor[] {
+    return JSON.parse(this.dataset.credentials);
+  }
+
+  get userChallenge(): string {
+    return this.dataset.userChallenge;
+  }
+
+  bindEvents() {
+    this.button.addEventListener('click', () => this.verify());
+  }
+
+  async verify() {
+    this.spinner.hidden = false;
+
+    const { userChallenge, credentials } = this;
+
+    try {
+      const result = await verifyWebauthnDevice({ userChallenge, credentials });
+      this.appendInput('credential_id', result.credentialId);
```

Comment:
> style question: the old code had these already in the DOM from the template and the JS wrote to them, is it more maintainable to dynamically create them in JS or just write the values?

Good question! I think that could work equally as well, and I don't really have a strong preference one way or the other. I think the mental model I had in mind here is to treat these as "just-in-time" values, since the inputs relevant for the server can vary depending on the path (error vs. success, for one).

Reply:
part of me thinks they would be better left in the HTML template, because then when we go to edit the controller to see what fields to read, we can check there instead of having to jump to JS? I'm not 100% sure

Reply:
Yeah, I can see the benefit of that. Let me see what I can do. The other consideration is how to implement the references on the JavaScript side, since following the existing pattern could get rather noisy with all the getter methods.

```js
get credentialIdInput() {
  return this.querySelector('[name=credential_id]');
}

// etc.
```

Can probably find some more elegant way to organize / reference them.

Reply:
I was tempted to get fancy with [`Proxy` objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy)...

```js
this.values = new Proxy({}, {
  get(_target, prop) {
    return this.querySelector(`[name="${prop}]`).value;
  },
  set(_target, prop, value) {
    this.querySelector(`[name="${prop}]`).value = value;
  }
});

// ...

this.values.credential_id = 'foo';
```

...but in retrospect, since we only need to set the value and not read it, maybe we can just keep something more like the existing `appendInput`, and adapt it to assume the input already exists.

```suggestion
      this.setInputValue('credential_id', result.credentialId);
```

```js
setInputValue(name, value) {
  this.querySelector(`[name="${name}"]`).value = value;
}
```

Reply:
Updated in 7194e13 to render inputs server-side.